exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -317,3 +334,238 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
 
 	return ctrl.Result{}, nil
 }
+
+// TODO: maybe add this to the phases list
+// reconcileMachinePoolMachines reconciles the Spec.InfrastructureRef object on a MachinePool.
+func (r *MachinePoolReconciler) reconcileMachinePoolMachines(ctx context.Context, mp *expv1.MachinePool, infraMachinePool *unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(2).Info("Reconciling MachinePool Machines")
+
+	var noKind bool
+	var noSelector bool
+
+	var infraMachineKind string
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineKind, "status", "infrastructureMachineKind"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noKind = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineKind from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+	log.V(2).Info("Got infraMachineKind from infra provider", "kind", infraMachineKind)
+
+	var infraMachineSelector metav1.LabelSelector
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineSelector, "status", "infrastructureMachineSelector"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noSelector = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineSelector from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+	log.V(2).Info("Got infraMachineSelector from infra provider", "selector", infraMachineSelector)
+
+	// If one missing but not both, return error
+	if noKind && noSelector {
+		// If proper MachinePool Machines aren't supported by the InfraMachinePool, just create Machines to match the replica count.
+		log.V(2).Info("No infraMachineKind or infraMachineSelector found, reconciling fallback machines")
+		return r.reconcileFallbackMachines(ctx, mp)
+	} else if noKind != noSelector {
+		log.V(2).Info("Only one of infraMachineKind or infraMachineSelector found, returning error")
+		return errors.Errorf("only one of infraMachineKind or infraMachineSelector found, both must be present")
+	} else {
```

Comment:
For better readability:

As there is nothing after the else anymore and the if cases above return.

```suggestion
	}
```

(Of course also remove the closing at the end)